chain/network/src/raw.rs

```diff
@@ -0,0 +1,345 @@
+        &mut self,
+        protocol_version: ProtocolVersion,
+        chain_id: &str,
+        genesis_hash: CryptoHash,
+        head_height: BlockHeight,
+    ) -> Result<(), ConnectError> {
+        let handshake = PeerMessage::Handshake(Handshake {
+            protocol_version,
+            oldest_supported_version: protocol_version - 2,
+            sender_peer_id: self.my_peer_id.clone(),
+            target_peer_id: self.peer_id.clone(),
+            // we have to set this even if we have no intention of listening since otherwise
+            // the peer will drop our connection
+            sender_listen_port: Some(24567),
+            sender_chain_info: PeerChainInfoV2 {
+                genesis_id: GenesisId { chain_id: chain_id.to_string(), hash: genesis_hash },
+                height: head_height,
+                tracked_shards: vec![0],
+                archival: false,
+            },
+            partial_edge_info: PartialEdgeInfo::new(
+                &self.my_peer_id,
+                &self.peer_id,
+                1,
+                &self.secret_key,
+            ),
+        });
+
+        self.write_message(&handshake).await?;
+
+        let start = Instant::now();
+
+        let (message, timestamp) = self.recv_message().await?;
+
+        match message {
+            // TODO: maybe check the handshake for sanity
+            PeerMessage::Handshake(_) => {
+                tracing::info!(target: "network", "handshake latency: {:?}", timestamp - start);
+            }
+            PeerMessage::HandshakeFailure(_peer_info, reason) => {
+                return Err(ConnectError::HandshakeFailure(reason))
+            }
+            _ => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "First message received from {:?} is not a handshake: {:?}",
+                        self, &message
+                    ),
+                )
+                .into())
+            }
+        };
+
+        Ok(())
+    }
+
+    async fn write_message(&mut self, msg: &PeerMessage) -> io::Result<()> {
+        let mut msg = msg.serialize(Encoding::Proto);
+        let mut buf = (msg.len() as u32).to_le_bytes().to_vec();
+        buf.append(&mut msg);
+        self.stream.write_all(&buf).await
+    }
+
+    // panics if there's not at least `len` + 4 bytes available
+    fn extract_msg(&mut self, len: usize) -> io::Result<PeerMessage> {
+        self.buf.advance(4);
+        let msg = PeerMessage::deserialize(Encoding::Proto, &self.buf[..len]);
+        self.buf.advance(len);
+        msg.map_err(|_| {
+            io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!("error parsing protobuf of length {}", len),
+            )
+        })
+    }
+
+    async fn do_read(&mut self) -> io::Result<()> {
+        let n = tokio::time::timeout(Duration::from_secs(5), self.stream.read_buf(&mut self.buf))
```

Comment:
hardcoding timeouts like this is generally not a good idea. How about you pass it as an argument? Ideally this future should be cancellable, but I guess we don't care that much about code used by probing.